src/Util/MigrationRollbackBatch.php

```diff
@@ -0,0 +1,288 @@
+    else {
+      throw new \Exception('Migration failed.');
+    }
+  }
+
+  /**
+   * Adds the migration rows to a queue
+   *
+   * @return int
+   */
+  private function enqueue(): int {
+    $this->iterator->rewind();
+
+    $this->queue = $this->getQueue();
+    $this->queue->deleteQueue();
+
+    while ($this->iterator->valid()) {
+      $this->queue->createItem($this->iterator->current());
+      $this->iterator->next();
+    }
+
+    return MigrationInterface::RESULT_COMPLETED;
+  }
+
+  /**
+   * Process each batch to roll back all contained rows.
+   *
+   * @param array $context
+   *   The batch context.
+   *
+   * @return int
+   *   MigrationInterface result constant
+   */
+  public function processBatch(array &$context): int {
+    try {
+      // Announce that rollback is about to happen.
+      $this->getEventDispatcher()->dispatch(new MigrateRollbackEvent($this->migration), MigrateEvents::PRE_ROLLBACK);
+
+      // Initialize class context var
+      $this->context = $context;
+      $context['finished'] = 0;
+
+      $context['message'] = $this->t('Processing of "@migration_id"', ['@migration_id' => $this->migration->id()]);
+
+      $result = $this->rollback();
+
+      $context['message'] = $this->t('"@migration_id" has been processed', ['@migration_id' => $this->migration->id()]);
+      $this->messenger->addMessage(
+        $this->t('"@migration_id" has been processed', ['@migration_id' => $this->migration->id()])
+      );
+
+      $context['results']['status'] = MigrationInterface::RESULT_COMPLETED;
+      $context['sandbox']['total'] = 0;
+      $context['finished'] = 1;
+
+      return $result;
+    }
+    catch (\Exception $e) {
+      $this->handleException($e, FALSE);
+      $this->messenger->addError(
+        $this->t(
+          'Rollback encountered error while processing batch: @e.', ['@e' => $e]
+        )
+      );
+
+      return MigrationInterface::RESULT_FAILED;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rollback() {
+    // Only begin the rollback operation if the migration is currently idle.
+    if ($this->migration->getStatus() !== MigrationInterface::STATUS_IDLE) {
+      $this->message->display($this->t(
+        'Migration @id is busy with another operation: @status',
+        ['@id' => $this->migration->id(), '@status' => $this->migration->getStatusLabel()]), 'error');
+      return MigrationInterface::RESULT_FAILED;
+    }
+
+    // Set the migration status to rolling back.
+    $this->migration->setStatus(MigrationInterface::STATUS_ROLLING_BACK);
+
+    $id_map = $this->getIdMap();
+    $destination = $this->migration->getDestinationPlugin();
+    $return = MigrationInterface::RESULT_COMPLETED;
+
+    while ($this->context['finished'] < 1) {
+      $item = $this->queue->claimItem();
+
+      if ($item->data === NULL) {
+        $this->message->display($this->t('Skipped processing due to null destination identifier.'));
+        $source_key = $id_map->currentSource();
```

Comment:
`$id_map->currentSource()` shouldn't necessarily be expected to be consistent with what item we pulled from the queue. It might be necessary to make the data in the queue more elaborate, tracking both the source and destination keys instead.